### PR TITLE
Stale proof verification

### PR DESCRIPTION
**Pull request by SoujanyaPonnapalli**

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1458,4 +1458,63 @@ export function VerifyWitness(
...
+  const oldNodeHashes = [];
+  for (const serializedOldNode of witness.proof) {
+    oldNodeHashes.push(hashAsBuffer(HashType.KECCAK256, serializedOldNode));
+  }
+  for (const [idx, witNode] of result.stack.entries()) {
+    const rlp = witNode.getRlpNodeEncoding(recentState.convertValue);
+    let recentHash = Buffer.from("");
+    if (rlp.length >= 32 || (idx === 0)) {
+      recentWitness.proof.push(rlp);
+      recentHash = hashAsBuffer(HashType.KECCAK256, rlp);
```
Comment (no2chem) [inline src/index.ts]: Do you need to hash the node again here? You should be able to retrieve the hash from the node (I guess you need to probably rebase against master).

File: src/index.ts [hunk 2]
```diff
@@ -1458,4 +1458,63 @@ export function VerifyWitness(
...
+    recentWitness.value = result.node.value;
+    const witProof: Buffer[] = [];
+    for (const [idx, witNode] of result.stack.entries()) {
+      const rlp = witNode.getRlpNodeEncoding(recentState.convertValue);
+      if (rlp.length >= 32 || (idx === 0)) {
+        witProof.push(rlp);
+      }
+    }
+    recentWitness.proof = witProof;
+    VerifyWitness(recentState.root, key, recentWitness);
```
[the remark this answers is not in the document]
Comment (SoujanyaPonnapalli) [reply src/index.ts]: Okay, this function need not return the recent witness either right?

[the remark this answers is not in the document]
Comment (no2chem) [reply src/index.ts]: Nope. Just throw an error if it fails

File: src/index.ts (hunk 1 above)
Comment (SoujanyaPonnapalli) [reply src/index.ts]: I need the hash to check if it matches with any of the previous node. And I also need to get the rlp encoding to construct the recent witness.

**Comment by SoujanyaPonnapalli** [comment]
For the cachedMerklePatriciaTree, should I take as a parameter:
- RlpWitness : {value: Buffer, proof: Buffer[]}? or a 
- Witness : {value: Buffer, proof: Array<MerklePatriciaTreeNode>} ?

File: src/index.ts [hunk 3]
```diff
@@ -638,7 +638,7 @@ export interface MerklePatriciaTreeOptions<K, V> {
 export class MerklePatriciaTree<K = Buffer, V = Buffer> implements
     MerkleTree<K, V> {
   /** The root node of the tree. */
-  private rootNode: MerklePatriciaTreeNode<V>;
+  rootNode: MerklePatriciaTreeNode<V>;
```
[the remark this answers is not in the document]
Comment (SoujanyaPonnapalli) [reply src/index.ts]: I might have to revert back to making it non-private for the cachedMerklePatriciaTree. I use it in the pruneCacheState() function.

File: src/index.ts [hunk 4]
```diff
@@ -1458,4 +1502,111 @@ export function VerifyWitness(
...
+        }
+        verifyWitness(recentState.root, key, curWit);
+        return;
+      }
+    }
+  }
+  throw new VerificationError("stale witness verification failed");
+}
+
+export class CachedMerklePatriciaTree extends MerklePatriciaTree<Buffer, Buffer> {
```
Comment (no2chem) [inline src/index.ts]: Should be typed, so 
```typescript
export class CachedMerklePatriciaTree<K,V> extends MerklePatriciaTree<K,V> {}
```

File: src/index.ts [hunk 5]
```diff
@@ -1458,4 +1502,111 @@ export function VerifyWitness(
...
+    super();
+    // Set the default maxCacheDepth to 6
+    this.maxCacheDepth = depth;
+  }
+
+  /** Returns the maxCacheDepth */
+  getmaxCacheDepth () : number {
+    return this.maxCacheDepth;
+  }
+
```
Comment (no2chem) [inline src/index.ts]: Where's the putWitness function we discussed? We want to take a Witness, verifying it against the cached tree. You can assume that the witness is RLP encoded.

File: src/index.ts [hunk 6]
```diff
@@ -1458,4 +1502,111 @@ export function VerifyWitness(
...
+  /**
+   * Recursively prunes the cachedStateTree to maxCacheDepth by adding HashNodes
+   * @param currNode current node at a depth, starting from rootNode at depth = 1
+   * @param depth Ranges from 1 to maxCacheDepth
+   */
+  pruneCacheState(currNode: MerklePatriciaTreeNode<Buffer> = this.rootNode, depth = 1) {
```
Comment (no2chem) [inline src/index.ts]: Looks good, but 
```typescript
MerklePatriciaTreeNode<V>
```

File: src/index.ts [hunk 7]
```diff
@@ -552,17 +557,55 @@ export class LeafNode<V> extends MerklePatriciaTreeNode<V> {
...
+export class HashNode<V> extends MerklePatriciaTreeNode<V> {
+  /** A hash node always has no nibbles. */
+  readonly nibbles = [];
+
+  /** The value of a hash node cannot be set. */
+  set value(val: V) {
+    throw new Error('Attempted to set the value of a NullNode');
+  }
+
+  /** nodeHash stores the hash of the next node. */
```
Comment (no2chem) [inline src/index.ts]: You mean the hash of this node, right?

File: src/index.ts (hunk 5 above)
Comment (SoujanyaPonnapalli) [reply src/index.ts]: Yeah I'm working on it, trying to test and fix it. Will be done in an hour or so.

Comment (SoujanyaPonnapalli) [reply src/index.ts]: I think putWitness shouldn't perform any verification, but just populate the cachedState right? And then we verify the stale witness using this populated cachedState.
We cannot really figure out if we want to add a witness to the cache or not just looking at the witness and the root right? We don't know if root is stale or more recent.

Comment (no2chem) [reply src/index.ts]: It should perform verification, otherwise it will add potentially incorrect data!

**Pull request merged**